Add: Debugging Interface (for review)

RaiVaibhav opened the pull request:
A debugging Interface Pull Request for review

RaiVaibhav commented:
This change has been done in the PR to minimize the failings the different tests.

RaiVaibhav commented on coalib/bearlib/__init__.py:
```diff
@@ -150,6 +156,7 @@ def wrapping_function(*args, **kwargs):
 
         wrapping_function.__metadata__ = new_metadata
 
+        wrapping_function.__doc__ = func.__doc__
```
This changes are needed, so that docstring of any bear can't get lost because of removing the the `@warps` tool

fneu replied:
okay. In cases like this it's even better to make a comment in the code since that will stay even when your github comment will not :)

Makman2 replied:
Why doesn't work `@wraps` again?

RaiVaibhav replied:
You want to know the reason why I remove the `@warps` form. If yes then because I want to check whether the `run()` method is using decorator or not, but how I am going to do that, so I removed the `@warp` tool so that I can check whether `decorators` is getting called or not by checking the name of `run()` method.

RaiVaibhav replied:
These name are used accordingly in creating a different cases,

RaiVaibhav commented on coalib/bearlib/aspects/decorators.py:
```diff
@@ -41,11 +43,16 @@ def _new_func(self, *args, **kwargs):
         # Keep metadata
         _new_func.__metadata__ = FunctionMetadata.from_function(func)
 
+        _new_func.__doc__ = func.__doc__
```
Same here

RaiVaibhav commented on coalib/bears/Bear.py:
```diff
@@ -354,7 +441,6 @@ def __json__(cls):
         """
         Override JSON export of ``Bear`` object.
         """
-        # json cannot serialize properties, so drop them
```
An unrelated change. I will revert it back

fneu commented on .moban.yaml:
```diff
@@ -34,6 +34,7 @@ dependencies:
   - setuptools>=21.0.0
```
first of all, separating your progress into multiple commits would be nicer to review, because I can then 'follow' your commits like a train of thought.

Now I basically have to understand everything you do at once.

I know that it's hard to get commits right when first writing new code, but if you haven't done so before, take a look at 'git rebase --interactive' to learn how to change previous commits in your branch.

RaiVaibhav replied:
I will do that :+1:

RaiVaibhav replied:
Done :+1:

fneu commented on coalib/bearlib/aspects/decorators.py:
```diff
@@ -4,7 +4,9 @@
-
+import pdb
+db = pdb.Pdb()
+db.prompt = '(coala-debugger)'
```
I'm not sure if we should change the prompt. People should recognize pdb and look up it's documentation. If they only see 'coala-debugger' they moght expect extensive documentation from our side?

RaiVaibhav replied:
Yes, you are right but change in color of string is needed :stuck_out_tongue:  :+1:

fneu commented on coalib/bears/Bear.py:
```diff
@@ -18,8 +19,26 @@
 from coalib.settings.FunctionMetadata import FunctionMetadata
 from coalib.settings.Section import Section
 from coalib.settings.ConfigurationGathering import get_config_directory
-
```
look up PEP8 order of imports ;)

RaiVaibhav replied:
Actually, my PR still needs the linting :stuck_out_tongue: , I haven't run `coala` on this change, was waiting for the review

fneu commented on coalib/bears/Bear.py:
```diff
@@ -274,6 +293,56 @@ def log_message(self, log_message, timestamp=None, **kwargs):
     def run(self, *args, dependency_results=None, **kwargs):
         raise NotImplementedError
 
+    def parenthesis_split(self, sentence, separator=',',
```
what is this needed for?

RaiVaibhav replied:
removed :+1:

fneu commented on coalib/bears/Bear.py:
```diff
@@ -288,6 +357,24 @@ def run_bear_from_section(self, args, kwargs):
             return
 
+        debug_bears = kwargs.get('debug_bears')
+        value_list = self.parenthesis_split(str(debug_bears))
```
I don't think this is necessary, why can't you interpret the setting as al list right away? look here: https://github.com/coala/coala/blob/master/coalib/settings/Setting.py#L84

RaiVaibhav replied:
Nice applying changes :+1:

RaiVaibhav commented on coalib/bears/Bear.py:
```diff
@@ -288,6 +357,24 @@ def run_bear_from_section(self, args, kwargs):
+        elif first_value_list=='true':
+            self.debugger_config(*args,**kwargs)
+            kwargs.pop('debug_bears')
+        elif str(self.__class__.__name__) in value_list:
```
this `condition` will be merge with above condition

RaiVaibhav commented on coalib/processes/Processing.py:
```diff
@@ -393,7 +404,8 @@ def instantiate_processes(section,
 
-    if debug:
+    if debug or debug_bears:
+        debug_bears = typed_list(str)(section['debug_bears'])
```
Used  `typed_list` and removed the `parenthesis_split` function. :+1:

RaiVaibhav replied:
Removing this already covered up there :+1:

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -587,6 +599,16 @@ def process_queues(processes,
                                Results were delivered to the user. Else False.
     """
+    try:
+        debug_bears = section['debug_bears'].value
```
you can invoke `bool`, then you don't have to compare against a string below

Makman2 replied:
Then you can completely avoid the `if` below

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -751,7 +773,18 @@ def execute_section(section,
+    try:
+        debug_bears = section['debug_bears'].value
+    except IndexError:
+        debug_bears = False
```
same as above

Makman2 commented on coalib/bears/Bear.py:
```diff
@@ -18,10 +22,27 @@
 from .meta import bearclass
 
 
+db = pdb.Pdb()
```
I would call it `debugger` ;)

Makman2 replied:
Or `dbg` at least

RaiVaibhav replied:
:P

Makman2 commented on coalib/bears/Bear.py:
```diff
@@ -18,10 +22,27 @@
+    try:
+        while True:
+            result = db.runcall(next, x)
+            results.append(result)
```
Didn't we want to call `db.runcall(func, *args, **kwargs)` above outside the loop as well?

RaiVaibhav replied:
Do you want me to step into decorators also?

Makman2 commented on coalib/bears/Bear.py:
```diff
@@ -274,6 +295,32 @@ def log_message(self, log_message, timestamp=None, **kwargs):
 
+    def debugger_config(self,*args,**kwargs):
+        debug_bears = kwargs.get('debug_bears')
+        kwargs.pop('debug_bears')
```
I believe you can do `debug_bears = kwargs.pop('debug_bears')`

Makman2 commented on coalib/bears/Bear.py:
```diff
@@ -274,6 +295,32 @@ def log_message(self, log_message, timestamp=None, **kwargs):
+    def debugger_config(self,*args,**kwargs):
+        debug_bears = kwargs.get('debug_bears')
+        kwargs.pop('debug_bears')
+        if ('linter' in str(self)) and \
```
I would rather rewrite `debugger_config` in `@linter` so it can implement its own debugger configuration. The `Bear` base class shouldn't know about `linter`.

RaiVaibhav replied:
I have to check where in `Linter` file I can call it :+1:

RaiVaibhav replied:
@Makman2  I think I have made one mistake here, check that case `1` and case `3` are calling same functions here, so 

IMO without case 1, it will do the same thing, so now Bear Base class will not know about linter :+1:

RaiVaibhav replied:
So writing the `debugger_config` in `@linter` will not be needed now.

Makman2 commented on coalib/bears/Bear.py:
```diff
@@ -288,6 +335,20 @@ def run_bear_from_section(self, args, kwargs):
+        else:
+            kwargs.pop('debug_bears')
+            logging.info("Passed the Bear name as a argument to debug")
+            return self.run(*args, **kwargs)
```
Instead of fiddling here with sections, it makes sense to make the `__init__` function accept a `debug` parameter, that controls debugging per bear. Then you would shift this logic to the bear-instantiation part of the core (then you check the section for debug configs, and instantiate bears accordingly in debug mode or not).

Makman2 replied:
`debug` is just a so essential feature, that it imo should reside in `__init__`^^ Would be also easier to port this to the NextGen-Core :+1:

RaiVaibhav replied:
Done till noon I will try to rewrite the `LInter` part then update the PR for review

Makman2 commented on coalib/bearlib/__init__.py:
```diff
@@ -7,8 +7,12 @@
 from coalib.settings.FunctionMetadata import FunctionMetadata
 
+db = pdb.Pdb()
+db.prompt = colorit.color_front('(pdb)',50,200,50)
```
Coloring shouldn't be done by default. I mean I would even prefer to have a powerline prompt here :P
Such design considerations should be shifted to the end of implementation, as this depends on your terminal used (This could really make problems with Windows)

Makman2 commented on .moban.yaml:
```diff
@@ -34,6 +34,7 @@ dependencies:
   - setuptools>=21.0.0
   - testfixtures~=5.3.1
   - unidiff~=0.5.2
+  - colorit
```
If possible let's rather avoid new packages that just deal with coloring ;)

RaiVaibhav commented:
Code extraction is in progress :P

Makman2 commented on coalib/bears/Bear.py:
```diff
@@ -274,6 +293,27 @@ def log_message(self, log_message, timestamp=None, **kwargs):
+            dbg.runcall(self.run,*args, **kwargs)
+            debug_mode_function(self.run,*args,**kwargs)
+
+        elif (isinstance(self.run(*args, **kwargs),(str,type(None)))) and \
```
You are rerunning `run` all the time, this is not okay. Run it once, and then check the return value.

Makman2 replied:
Hm you want to get the return type by running it... this is not a good design. The problem is we can't predict the type without running the bear, so you have to find something different.

RaiVaibhav committed:
DefaultArgParser.py: Add debug bears argument

Add --debug-bears argument to coala

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -303,7 +309,15 @@ def filter_raising_callables(it, exception, *args, debug=False, **kwargs):
+                ((elem.__name__) in debug_bears):
+                kwargs['debug_flag'] = True
+            else:
+                kwargs['debug_flag'] = False
```
you can compress this `if` to a ternary `if`, since two of the branches are the same.

RaiVaibhav replied:
My main reason here to spilt into three because if I do something like
```
if ((str(debug_bears[0]).lower().strip())=='true') or (
                (elem.__name__) in debug_bears):
    kwargs['debug_flag'] = True
else:
    kwargs['debug_flag'] = False
```
then for the case where `debug_bears` keys is `False` it will throw the error as `debug_bears` is not a list (`bool type is not subscriptable`)
or second way is to add one more check that whether type of `debug_bears` is bool or not
and after that make simply `if-else`

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -303,7 +309,15 @@ def filter_raising_callables(it, exception, *args, debug=False, **kwargs):
         try:
+            if debug_bears is False:
+                kwargs['debug_flag'] = False
+            elif ((str(debug_bears[0]).lower().strip())=='true') or \
```
please avoid `\` for line continuations :P

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -303,7 +309,15 @@ def filter_raising_callables(it, exception, *args, debug=False, **kwargs):
+            else:
+                kwargs['debug_flag'] = False
             yield elem(*args, **kwargs)
+            kwargs.pop('debug_flag')
```
just leave it there ;)

RaiVaibhav replied:
~~@Makman2 this change is needed, beacuse of iterator during second  iteration it will contain a `debug_flag` and will through error during instantiation of second bear,~~

~~Please correct me if I misunderstood your comment~~

RaiVaibhav replied:
It's working because it's overwriting the `debug_flag` `kwarg`.

RaiVaibhav replied:
I think I was making some mistake previously then :+1:

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -291,7 +292,12 @@ def get_file_dict(filename_list, log_printer=None, allow_raw_files=False):
+                             *args,
+                             debug=False,
+                             debug_bears=False,
+                             **kwargs):
```
do you really need to modify this function? This one is very generic and has the only purpose to filter out return values, whose function that produces them raises an error.

RaiVaibhav replied:
This was actually a reason because of which I wanted to review this part of code, because this is where Bear is getting instantiated and also current test of function is thing which I can't make it to work because it will get unexpected kwarg `debug_flag`

Makman2 replied:
I mean pass it via the already existing `args` parameter ;)

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -380,6 +397,12 @@ def instantiate_processes(section,
+        debug_bears = False if section['debug_bears'].value=='False' else (
+            section['debug_bears'].value)
+    except IndexError:
+        debug_bears = False
```
maybe do an `'debug_bears' in section` inside the above `if` expression? Would compactify code a bit more. Don't worry: Due to Python's short circuitance following conditionals won't be executed if `'debug_bears' in section` evaluates to `False`

RaiVaibhav replied:
I think this will be best ?

```python
debug_bears = False if section['debug_bears'].value=='False' else (
        typed_list(str)(section['debug_bears'].value))
```

RaiVaibhav replied:
This will remove the extra line 401 also

Makman2 replied:
You are msising the `'debug_bears' in section` check, otherwise there might be lookup errors if the key doesn't exists:
```python
debug_bears = (False
               if 'debug_bears' in section and section['debug_bears'].value == 'False' else
               typed_list(str)(section['debug_bears'].value))
```

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -439,7 +464,6 @@ def instantiate_processes(section,
     file_dict = {filename: complete_file_dict[filename]
                  for filename in filename_list
                  if filename in complete_file_dict}
-
```
Obsolete change

fneu commented on coalib/bears/Bear.py:
```diff
@@ -18,10 +21,30 @@
+
+dbg = db()
+dbg.do_q = dbg.do_continue
+dbg.do_c = dbg.do_continue
```
remove this?

Makman2 commented on coalib/bears/Bear.py:
```diff
@@ -232,7 +254,8 @@ def maintainers_emails(cls):
                  message_queue,
-                 timeout=0):
+                 timeout=0,
+                 debug_flag=False):
```
Just call it `debug`, the `_flag` part is redundant, any boolean argument can be called a "flag" ;)

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -587,6 +591,11 @@ def process_queues(processes,
+    debug_bears = (False
+                   if 'debug_bears' not in section or (
+                   section['debug_bears'].value == 'False') else (
+                   typed_list(str)(section['debug_bears'])))
```
You are doing it too often imo. There should be one location to extract the actual `debug` setting from the section, and then pass it down via parameters.

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -329,27 +322,32 @@ def instantiate_bears(section,
+                                                     message_queue,
+                                                     timeout=0.1,
+                                                     debug_flag=debug_flag))
+        except:
```
No broad except

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -290,32 +291,24 @@ def get_file_dict(filename_list, log_printer=None, allow_raw_files=False):
+        debug_flag = True
+    else:
+        debug_flag = False
+    return debug_flag
```
Do you need this whole stuff? I never see you using those string-branches^^

RaiVaibhav replied:
Yes I needed those, previously it was included in the `filter_raisable_callable_function` and because of this function the `debug_flag` will set to `True` to only those bears which are passed to `--debug-bears`

RaiVaibhav replied:
With the help of [test cases](https://github.com/RaiVaibhav/coala/pull/5/files#diff-7c16cc0dc5d9e615bbb54239955f7b5eR609), I hoping that I am clearing the point

Makman2 commented on coalib/processes/Processing.py:
```diff
@@ -329,27 +322,32 @@ def instantiate_bears(section,
+            instantiated_local_bear_list.append(bear(section,
+                                                     message_queue,
+                                                     timeout=0.1,
+                                                     debug_flag=debug_flag))
```
~~Imo you don't have to switch to a manual-appending-strategy, you can still use the generator-syntax from above^^~~

Makman2 replied:
Nvm

RaiVaibhav committed:
Processing.py: Add debug bears process

Enable single process so that pdb
can work in single process environment
for debugging the bears and initialize
each Bear with `debugger` flag, this flag
will be use to enable the Debugger
Interface.

RaiVaibhav committed:
GlobalBear.py: Add `debugger` flag

RaiVaibhav committed:
Bears.py: Add the Debbuging Interface

Perform Debugging on the `run()` method of
Bear.

RaiVaibhav committed:
tests: Add Debugging interface test

Added test in `BearTest.py` and
`ProcessingTest.py`.

RaiVaibhav commented:
Closing moved to https://github.com/coala/coala/pull/5511
